### ISSUE TITLE
A case for `prefer-flat-map`

### ISSUE THREAD

**Issue by tinovyatkin** [sindresorhus/eslint-plugin-unicorn#311]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/311)

I think it will be way greater rule if it will be able to identify legacy (before `.flat` got supported) code and suggest to refactor to `flatMap`. I believe one of the most common scenario was use of `[].concat` with array spread:

```js
[].concat(...someArray.map(fn))
// is the same
someArray.flatMap(fn)
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Sponsors (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by sindresorhus** [comment]
Good idea!

**Comment by fisker** [comment]
should we consider `Array#@@isConcatSpreadable`?

**Comment by sindresorhus** [comment]
@fisker Consider it how? Can you elaborate?

**Comment by fisker** [comment]
@sindresorhus 

check this https://jsbin.com/farogop/edit?js,console 

thought this maybe not very common case,
I'm just saying, It's possible not the same as flatMap,  and we can't guarantee new Objects not ConcatSpreadable never come in future

**Comment by MrHen** [comment]
@fisker I'm comfortable using `// eslint-disable-next-line prefer-flat-map` to handle cases like that.

**Comment by sindresorhus** [comment]
That's too much of an obscure use-case to handle.

**Comment by fisker** [comment]
Imaging this `fn` is from a third-party library, which is `isConcatSpreadable` false

```js
[].concat(...someArray.map(fn))
```

It works as expected, when you wrote and tested, you may not think about adding `// eslint-disable-next-line prefer-flat-map`.
When you format your code and publish, It breaks. You have to debug your program to locate the problem.

If a linter can't be safe, what's the point of using it.

I insist not to fix this code, at least an option, people should know this may cause problems

**Comment by sindresorhus** [comment]
@fisker This seems very unlikely, but sure, we can omit fixing this case.

> If a linter can't be safe, what's the point of using it.

Then you should probably not use JavaScript. Most lint rules are not 100% safe. That is just impossible for such a loose language as JS.

**Comment by IssueHuntBot** [comment]
[@issuehunt](https://issuehunt.io/sponsors) has funded $40.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/311) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/311).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.
